core/src/test/scala/org/http4s/client/jdkhttpclient/JdkWSClientSpec.scala

```diff
@@ -0,0 +1,183 @@
+            WSFrame.Text("foo"),
+            WSFrame.Text("bar")
+          )
+        )
+    }
+
+    "group frames by their `last` attribute in high-level mode" in {
+      webSocket
+        .connectHighLevel(WSRequest(wsUri))
+        .use { conn =>
+          for {
+            _ <- conn.sendMany(
+              List(
+                WSFrame.Text("1", last = false),
+                WSFrame.Text("2", last = false),
+                WSFrame.Text("3"),
+                WSFrame.Binary(ByteVector(1)),
+                WSFrame.Binary(ByteVector(2), last = false),
+                WSFrame.Binary(ByteVector(3), last = false),
+                WSFrame.Binary(ByteVector(4)),
+                WSFrame.Text("4", last = false),
+                WSFrame.Text("5"),
+                WSFrame.Binary(ByteVector(5), last = false),
+                WSFrame.Binary(ByteVector(6)),
+                WSFrame.Text("6"),
+                WSFrame.Binary(ByteVector(7), last = false)
+              )
+            )
+            _ <- conn.sendClose()
+            recv <- conn.receiveStream.compile.toList
+          } yield recv
+        }
+        .map(
+          _ mustEqual List(
+            WSFrame.Text("123"),
+            WSFrame.Binary(ByteVector(1)),
+            WSFrame.Binary(ByteVector(2, 3, 4)),
+            WSFrame.Text("45"),
+            WSFrame.Binary(ByteVector(5, 6)),
+            WSFrame.Text("6")
+          )
+        )
+    }
+
+    "automatically close the connection" in {
+      for {
+        ref <- Ref[IO].of(List.empty[WebSocketFrame])
+        finished <- Deferred[IO, Unit]
+        // we use Java-Websocket because Blaze has a bug concerning the handling of Close frames and shutting down
+        server = new WebSocketServer(new InetSocketAddress("localhost", 8080)) {
+          override def onOpen(conn: WebSocket, handshake: ClientHandshake) = ()
+          override def onClose(conn: WebSocket, code: Int, reason: String, remote: Boolean) =
+            ref
+              .update(_ :+ WebSocketFrame.Close(code, reason).fold(throw _, identity))
+              .unsafeRunSync()
+          override def onMessage(conn: WebSocket, message: String) =
+            ref.update(_ :+ WebSocketFrame.Text(message)).unsafeRunSync()
+          override def onMessage(conn: WebSocket, message: ByteBuffer) =
+            ref.update(_ :+ WebSocketFrame.Binary(ByteVector(message))).unsafeRunSync()
+          override def onError(conn: WebSocket, ex: Exception) = println(s"WS error $ex")
+          override def onStart() = {
+            val req = WSRequest(uri"ws://localhost:8080")
+            val p = for {
+              _ <- webSocket.connect(req).use { conn =>
+                conn.send(WSFrame.Text("hi blaze"))
+              }
+              _ <- Timer[IO].sleep(1.seconds)
+              _ <- webSocket.connectHighLevel(req).use { conn =>
+                conn.send(WSFrame.Text("hey blaze"))
+              }
+              _ <- Timer[IO].sleep(1.seconds)
```

Comment:
The `expect` that I did on my branch means we can be more optimistic and finish quicker.  I'll probably port that over in a follow up.  At this point, I'd like to get this landed and get people trying it because it works.

Reply:
Ah yes, I always forget that matchers are impure...